frontend/src/components/elements/PlotlyChart/CustomTheme.tsx

```diff
@@ -0,0 +1,630 @@
+    ) {
+      graph.customdata.forEach((data: any, dataIndex: any) => {
+        const dataString = data.toString()
+        if (Array.isArray(data) && data.length > 0) {
+          if (customDataToDataIdx.has(dataString)) {
+            customDataToDataIdx.set(
+              dataString,
+              // @ts-ignore
+              customDataToDataIdx.get(dataString)?.concat(dataIndex)
+            )
+          } else {
+            customDataToDataIdx.set(dataString, [dataIndex])
+          }
+        }
+      })
+      graphIdxToCustomData.set(graphIndex, customDataToDataIdx)
+    }
+    if (graph.legendgroup !== undefined) {
+      if (legendGroupToIndexes.has(graph.legendgroup)) {
+        legendGroupToIndexes.set(
+          graph.legendgroup,
+          // @ts-ignore
+          legendGroupToIndexes.get(graph.legendgroup).concat(graphIndex)
+        )
+      } else {
+        legendGroupToIndexes.set(graph.legendgroup, [graphIndex])
+      }
+    }
+  })
+
+  let colorIndex = 0
+  legendGroupToIndexes.forEach((dataIdx: number[]) => {
+    dataIdx.forEach((index: number) => {
+      if (data[index].line !== undefined) {
+        // dont assign colors for dist plot boxes when they're transparent
+        if (
+          data[index].type !== "box" &&
+          data[index].line.color !== "rgba(255,255,255,0)" &&
+          data[index].line.color !== "transparent"
+        ) {
+          data[index].line = assign(data[index].line, {
+            color: categoryColors[colorIndex % categoryColors.length],
+          })
+        }
+      }
+      if (
+        data[index].marker !== undefined &&
+        typeof data[index].marker.color === "string"
+      ) {
+        data[index].marker = assign(data[index].marker, {
+          color: categoryColors[colorIndex % categoryColors.length],
+        })
+      }
+    })
+    colorIndex++
+  })
+
+  colorIndex = 0
+  graphIdxToCustomData.forEach(
+    (customData: Map<string, number[]>, dataIndex: number) => {
+      customData.forEach(markerIndexes => {
+        markerIndexes.forEach(markerIndex => {
+          data[dataIndex].marker.colors[markerIndex] =
+            categoryColors[colorIndex % categoryColors.length]
+        })
+        colorIndex++
+      })
+    }
+  )
+}
+
+/**
+ * This overrides the colorscale (continuous colorscale) to all graphs.
+ * @param data - spec.data
+ */
+export function applyColorscale(data: any, theme: Theme): any {
+  data.forEach((entry: any) => {
+    entry = assign(entry, {
+      colorscale: hasLightBackgroundColor(theme)
+        ? sequentialColorscaleLightTheme
+        : sequentialColorscaleDarkTheme,
+    })
+  })
+  return data
+}
+
+/**
+ * This applies colors specifically for the table, candlestick, and waterfall plot
+ * because their dictionary structure is different from other more regular charts.
+ * @param data - spec.data
+ */
+export function applyUniqueGraphColorsData(data: any, theme: Theme): void {
```

Comment:
Are the charts in the e2e tests covering all the logic in those methods: `applyUniqueGraphColorsData`, `applyColorscale`, `changeDiscreteColors`, `applyStreamlitThemeTemplateData`, `applyStreamlitThemeData`? We should be able to detect it via our e2e tests (or potentially jest tests) in case anything within those methods breaks based on changes to plotly.

Reply:
Writing here all the tests that are in our e2e script:

ApplyUniqueGraphColorsData: Table Plot, Candlestick Chart, Waterfall Chart

applyColorscale: contour will definitely cover this.

applyDiscreteColors: Need dist plot and plot with customData (sunburst chart)

applyStreamlitThemeTemplateLayout: Bar charts and contour plot would cover this method.

applyStreamlitThemeTemplateLayout: Ternary plots are needed specifically for this method.

applyStreamlitThemeData: Bubble charts needed since we remove the marker.

Through the above, it looks like these would be a good combination for everything:
Table Plot, Candlestick Chart, Waterfall Chart, contour plot, dist plot, sunburst chart, bar chart, ternary plot

for jest tests, we would just be checking properties so I don't think they're really valid tests and looking for what we want.

Reply:
Sounds good, let's add these tests to e2e 👍

Reply:
Maybe with a small comment in the e2e tests which logic they are covering